// pver/dist/cli.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// cli.ts
var import_yargs = __toESM(require("yargs"));
var import_helpers = require("yargs/helpers");

// src/app-context/index.ts
var import_promises = __toESM(require("fs/promises"));
var getAppContext = async ({
  argv,
  has_package_json,
  has_git_dir
}) => {
  const release_methods = argv.release_methods ?? [];
  const readme_files = [];
  if (argv.git) release_methods.push("git");
  if (argv.npm) release_methods.push("npm");
  if (argv.pushMain) release_methods.push("push-main");
  if (argv.readme) release_methods.push("readme");
  if (argv.packageJson) release_methods.push("package-json");
  if (argv.mdfile) {
    const mdfiles = Array.isArray(argv.mdfile) ? argv.mdfile : [argv.mdfile];
    for (const file of mdfiles) {
      if (typeof file === "string" && file.trim().length > 0) {
        readme_files.push(file);
      }
    }
  }
  if (argv.readme && readme_files.length === 0) {
    readme_files.push("README.md", "README.txt", "readme.md", "readme.txt");
  }
  if (has_git_dir === void 0) {
    has_git_dir = await import_promises.default.stat(".git").then(() => true).catch((e) => false);
    has_package_json = await import_promises.default.stat("package.json").then(() => true).catch((e) => false);
  }
  if (release_methods.length === 0) {
    if (has_git_dir) {
      release_methods.push("git");
      if (argv.pushMain !== false) {
        release_methods.push("push-main");
      }
    }
    if (has_package_json) release_methods.push("npm");
  }
  return {
    current_directory: process.cwd(),
    current_method: argv.current ?? "auto",
    transition_method: argv.transition ?? "auto",
    release_methods: [...new Set(release_methods)],
    readme_files: [...new Set(readme_files)]
  };
};

// src/analyze/current/get-current-package-json-version.ts
var import_node_fs = __toESM(require("fs"));
var getCurrentPackageJsonVersion = (ctx) => {
  const package_json_path = `${ctx.current_directory}/package.json`;
  try {
    const packageJson = JSON.parse(
      import_node_fs.default.readFileSync(package_json_path).toString()
    );
    return packageJson.version;
  } catch (e) {
    throw new Error(
      `Could not read package.json at ${package_json_path}: ${e.toString()}`
    );
  }
};

// src/analyze/current/get-last-commit-altering-package-json-version.ts
var import_simple_git = __toESM(require("simple-git"));
var getLastCommitAlteringPackageJsonVersion = async (ctx) => {
  const git = (0, import_simple_git.default)(ctx.current_directory);
  let log = await git.log({ file: "package.json" });
  if (log.all.length === 1) {
    await git.fetch(["origin", "main", "--depth=30"]);
    log = await git.log({ file: "package.json" });
  }
  console.log(`Detected ${log.all.length} commits altering package.json`);
  if (!log.latest) return null;
  for (const commit of log.all) {
    const diff = await git.show(["-U0", commit.hash, "--", "package.json"]);
    if (diff.includes('"version":')) return commit.hash;
  }
  return null;
};

// src/analyze/transition/get-simple-git-transitions.ts
var import_simple_git2 = require("simple-git");

// src/analyze/transition/get-next-version.ts
var getNextVersion = (current_version, change) => {
  const [version, tag = ""] = current_version.split("-");
  const [major, minor, patch] = version.split(".").map((v) => parseInt(v));
  const suffix = tag ? `-${tag}` : "";
  if (change === "bigrelease") return `${major + 1}.0.0${suffix}`;
  if (change === "announce") return `${major}.${minor + 1}.0${suffix}`;
  if (change === "increment") return `${major}.${minor}.${patch + 1}${suffix}`;
  throw new Error(`Invalid change type: "${change}"`);
};

// src/analyze/transition/get-simple-git-transitions.ts
var getSimpleGitTransitions = async ({
  current_version,
  current_version_commit_sha
}, ctx) => {
  const git = (0, import_simple_git2.simpleGit)(ctx.current_directory);
  const log = await git.log();
  if (!log.latest) throw new Error("No git history (commit something!)");
  if (log.latest.message.includes("[norelease]")) {
    console.log("Last commit contains [norelease], skipping version increment");
    return current_version;
  }
  if (log.latest && log.latest.hash === current_version_commit_sha)
    return current_version;
  const logs_since_current = [];
  for (const commit of log.all) {
    if (commit.hash === current_version_commit_sha) break;
    logs_since_current.push(commit);
  }
  let change = "increment";
  if (logs_since_current.some(
    (c) => c.message.toLowerCase().startsWith("bigrelease:")
  )) {
    change = "bigrelease";
  } else if (logs_since_current.some(
    (c) => c.message.toLowerCase().startsWith("announce:")
  )) {
    change = "announce";
  }
  return getNextVersion(current_version, change);
};

// src/analyze/current/check-if-git-tag-exists-for-version.ts
var import_simple_git3 = __toESM(require("simple-git"));
var checkIfGitTagExistsForVersion = async (ctx, version) => {
  const git = (0, import_simple_git3.default)(ctx.current_directory);
  await git.fetch(["--tags"]);
  const tags = await git.tags();
  return tags.all.includes(`v${version}`);
};

// src/analyze/index.ts
var analyze = async (ctx) => {
  let current_method = ctx.current_method;
  if (current_method === "auto") current_method = "package.json";
  let transition_method = ctx.transition_method;
  if (transition_method === "auto") transition_method = "simplegit";
  let current_version, current_version_commit_sha = null;
  if (current_method === "package.json") {
    current_version = await getCurrentPackageJsonVersion(ctx);
    if (transition_method === "simplegit")
      current_version_commit_sha = await getLastCommitAlteringPackageJsonVersion(ctx);
  }
  if (!current_version)
    throw new Error(
      `Couldn't find current version using --current-method: ${ctx.current_method}[${current_method}]`
    );
  let next_version;
  if (transition_method === "simplegit") {
    next_version = await getSimpleGitTransitions(
      {
        current_version,
        current_version_commit_sha
      },
      ctx
    );
    console.log(`current version: ${current_version}`);
    console.log(`candidate next version: ${next_version}`);
    if (next_version === current_version) {
      console.log("No changes detected, exiting");
      process.exit(0);
    }
    if (next_version !== current_version) {
      for (let i = 0; i < 10; i++) {
        if (!await checkIfGitTagExistsForVersion(ctx, next_version)) break;
        console.log(
          `Next version already exists as a git tag: ${next_version}. Incrementing version by an increment.`
        );
        next_version = getNextVersion(next_version, "increment");
      }
    }
    if (await checkIfGitTagExistsForVersion(ctx, next_version))
      throw new Error(
        `Next version already exists as a git tag: ${next_version}. We tried to incremenet 10 times but the tag still exists. Consider deleting some git tags.`
      );
  }
  if (!next_version)
    throw new Error(
      `Couldn't find next version using --transition-method: ${ctx.transition_method}[${transition_method}]`
    );
  return {
    current_method,
    transition_method,
    current_version,
    next_version
  };
};

// src/stage/make-git-tag.ts
var import_simple_git4 = __toESM(require("simple-git"));
var makeGitTag = async (tag, ctx) => {
  console.log(`Creating git tag ${tag}`);
  await (0, import_simple_git4.default)(ctx.current_directory).addTag(tag).catch((e) => {
    if (e.toString().includes("already exists")) {
      console.log(`Tag ${tag} already exists, skipping`);
      return;
    }
    throw e;
  });
};

// src/stage/update-package-json.ts
var import_package_json = __toESM(require("@npmcli/package-json"));
var updatePackageJson = async (new_version, ctx) => {
  console.log(`Updating package.json to version ${new_version}`);
  const pkg_json = await import_package_json.default.load(ctx.current_directory);
  pkg_json.update({ version: new_version });
  await pkg_json.save();
};

// src/stage/add-commit-changes.ts
var import_simple_git5 = __toESM(require("simple-git"));
var addCommitChanges = async (new_version, files_to_add, ctx) => {
  const files = [...new Set(files_to_add)];
  if (files.length === 0) {
    console.log("No file changes to commit, skipping commit step");
    return;
  }
  console.log("Adding files and committing changes");
  const git = (0, import_simple_git5.default)(ctx.current_directory);
  await git.add(files);
  await git.commit(`v${new_version}`);
};

// src/stage/make-sure-git-configured.ts
var import_simple_git6 = __toESM(require("simple-git"));
var makeSureGitConfigured = async (ctx) => {
  console.log("Making sure git is configured...");
  const git = (0, import_simple_git6.default)(ctx.current_directory);
  const user_email = process.env.GIT_USER_EMAIL || (process.env.GITHUB_ACTIONS ? "actions@github.com" : void 0);
  const user_name = process.env.GIT_USER_NAME || (process.env.GITHUB_ACTIONS ? "GitHub Actions" : void 0);
  if (user_email) {
    await git.addConfig("user.email", user_email, void 0, "local");
  }
  if (user_name) {
    await git.addConfig("user.name", user_name, void 0, "local");
  }
};

// src/stage/update-readme.ts
var import_promises2 = __toESM(require("fs/promises"));
var import_path = __toESM(require("path"));
var escapeRegExp = (input) => {
  const specials = new Set(".*+?^${}()|[]\\".split(""));
  let escaped = "";
  for (const char of input) {
    if (specials.has(char)) {
      escaped += `\\${char}`;
    } else {
      escaped += char;
    }
  }
  return escaped;
};
var updateReadme = async (current_version, next_version, ctx) => {
  const updated_files = [];
  if (ctx.readme_files.length === 0) return updated_files;
  const root = ctx.current_directory ?? process.cwd();
  const replacements = [
    [new RegExp(escapeRegExp(`v${current_version}`), "g"), `v${next_version}`],
    [new RegExp(escapeRegExp(current_version), "g"), next_version]
  ];
  const missing_files = [];
  const files_without_version = [];
  for (const file of ctx.readme_files) {
    const absolute_path = import_path.default.isAbsolute(file) ? file : import_path.default.join(root, file);
    let original;
    try {
      original = await import_promises2.default.readFile(absolute_path, "utf8");
    } catch (error) {
      if (error && error.code === "ENOENT") {
        missing_files.push(file);
        continue;
      }
      throw new Error(`Failed to read README file ${file}: ${error}`);
    }
    let updated = original;
    for (const [pattern, replacement] of replacements) {
      updated = updated.replace(pattern, replacement);
    }
    if (updated === original) {
      files_without_version.push(file);
      continue;
    }
    await import_promises2.default.writeFile(absolute_path, updated);
    const relative_path = import_path.default.relative(root, absolute_path) || import_path.default.basename(file);
    updated_files.push(relative_path);
  }
  if (updated_files.length === 0) {
    if (ctx.readme_files.length > 0) {
      if (files_without_version.length > 0) {
        throw new Error(
          `None of the specified README files contained version ${current_version}. Checked: ${files_without_version.join(", ")}`
        );
      }
      if (missing_files.length === ctx.readme_files.length) {
        throw new Error(
          `Unable to find any README files to update. Checked: ${ctx.readme_files.join(", ")}`
        );
      }
    }
  }
  return updated_files;
};

// src/stage/index.ts
var stage = async (ctx) => {
  const analysis = await analyze(ctx);
  console.log(`current version: ${analysis.current_version}`);
  console.log(`next version: ${analysis.next_version}`);
  if (analysis.current_version === analysis.next_version) {
    throw new Error(
      `Next version is the same as current version, not releasing. Check the transition method if you expected a version bump`
    );
  }
  console.log("");
  console.log(`Using staging methods: ${ctx.release_methods.join(", ")}`);
  await makeSureGitConfigured(ctx);
  const files_to_add = [];
  if (ctx.release_methods.includes("git")) {
    await makeGitTag(`v${analysis.next_version}`, ctx);
  }
  if (ctx.release_methods.includes("npm") || ctx.release_methods.includes("package-json")) {
    await updatePackageJson(analysis.next_version, ctx);
    files_to_add.push("package.json");
  }
  if (ctx.release_methods.includes("readme")) {
    const readme_files = await updateReadme(
      analysis.current_version,
      analysis.next_version,
      ctx
    );
    files_to_add.push(...readme_files);
  }
  if (files_to_add.length > 0) {
    await addCommitChanges(analysis.next_version, files_to_add, ctx);
  }
};

// src/release/npm-publish.ts
var import_child_process = require("child_process");
var npmPublish = (ctx) => {
  const result = (0, import_child_process.spawnSync)("npm", ["publish"], {
    cwd: ctx.current_directory,
    env: {
      ...process.env,
      NODE_AUTH_TOKEN: process.env.NODE_AUTH_TOKEN || process.env.NPM_TOKEN
    },
    stdio: "inherit"
  });
  if (result.status !== 0) {
    throw new Error("npm publish failed");
  }
};

// src/release/push-git-tag.ts
var import_simple_git7 = __toESM(require("simple-git"));

// src/release/configure-origin.ts
var normalizeGithubRemoteUrl = (remoteUrl) => {
  const trimmedUrl = remoteUrl.trim();
  const sshMatch = trimmedUrl.match(/^git@github\.com:(.+)$/);
  if (sshMatch) {
    return `https://github.com/${sshMatch[1]}`;
  }
  const sshProtocolMatch = trimmedUrl.match(/^ssh:\/\/git@github\.com\/(.+)$/);
  if (sshProtocolMatch) {
    return `https://github.com/${sshProtocolMatch[1]}`;
  }
  return trimmedUrl;
};
var addGithubTokenToRemoteUrl = (remoteUrl, token) => {
  const normalizedUrl = normalizeGithubRemoteUrl(remoteUrl);
  const url = new URL(normalizedUrl);
  if (url.hostname !== "github.com") {
    return normalizedUrl;
  }
  url.username = "x-access-token";
  url.password = token.trim();
  return url.toString().replace(/\/$/, "");
};
var configureOrigin = async (git) => {
  const githubToken = process.env.PVER_GITHUB_TOKEN || process.env.GITHUB_TOKEN;
  if (!githubToken) return;
  const remoteUrl = await git.remote(["get-url", "origin"]);
  const newUrl = addGithubTokenToRemoteUrl(remoteUrl, githubToken);
  if (newUrl === remoteUrl.trim()) return;
  await git.removeRemote("origin");
  await git.addRemote("origin", newUrl);
};

// src/release/push-git-tag.ts
var pushGitTag = async (tag, ctx) => {
  console.log("Pushing git tag", tag);
  let git = (0, import_simple_git7.default)(ctx.current_directory);
  await configureOrigin(git);
  await git.push("origin", tag);
};

// src/release/push-to-main.ts
var import_simple_git8 = __toESM(require("simple-git"));
var pushToMain = async (ctx) => {
  console.log("Pushing to main");
  let git = (0, import_simple_git8.default)(ctx.current_directory);
  await configureOrigin(git);
  await git.push(["origin", `HEAD:${process.env.GITHUB_REF ?? "main"}`]);
};

// src/release/sync-with-remote.ts
var import_simple_git9 = __toESM(require("simple-git"));
var syncWithRemote = async (ctx) => {
  console.log("Syncing with remote main");
  let git = (0, import_simple_git9.default)(ctx.current_directory);
  await configureOrigin(git);
  try {
    await git.pull("origin", "main", ["--rebase"]);
  } catch (e) {
    if (e.toString().includes("unstaged changes")) {
      const status = await git.status();
      const changed_files = [
        ...status.not_added,
        ...status.created,
        ...status.deleted,
        ...status.modified,
        ...status.renamed.map((r) => r.to)
      ];
      if (changed_files.length) {
        console.error(
          "Cannot pull with rebase due to unstaged changes in the following files:"
        );
        for (const file of changed_files) {
          console.error(" -", file);
        }
      }
    }
    throw e;
  }
};

// src/release/index.ts
var release = async (ctx) => {
  const analysis = await analyze(ctx);
  await stage(ctx);
  console.log(`Release methods enabled: ${ctx.release_methods.join(",")}`);
  await syncWithRemote(ctx);
  if (ctx.release_methods.includes("git")) {
    await pushGitTag(`v${analysis.next_version}`, ctx);
  }
  if (ctx.release_methods.includes("npm")) {
    await npmPublish(ctx);
  }
  if (ctx.release_methods.includes("push-main")) {
    await pushToMain(ctx);
  }
};

// cli.ts
if (process.env.NPM_TOKEN && !process.env.NODE_AUTH_TOKEN) {
  process.env.NODE_AUTH_TOKEN = process.env.NPM_TOKEN;
}
var handleCommandError = (e) => {
  console.error(e.stack);
  console.error(e.toString());
  process.exit(1);
};
(0, import_yargs.default)((0, import_helpers.hideBin)(process.argv)).command(
  "analyze",
  "Automatically compute the latest Pragmatic Version using the git history",
  (yargs2) => {
    return yargs2.option("current", {
      describe: "Method to compute the current version",
      choices: ["auto", "package.json"],
      default: "auto"
    }).option("transition", {
      describe: "Method to compute next version(s) from current version",
      choices: ["auto", "simplegit"],
      default: "auto"
    });
  },
  async (argv) => {
    const analysis = await analyze(await getAppContext({ argv })).catch(
      handleCommandError
    );
    console.log(analysis);
  }
).command(
  "release [action]",
  "Automatically compute and release a new version",
  (yargs2) => {
    return yargs2.positional("action", {
      describe: "The release action to perform",
      default: "auto",
      choices: ["auto", "increment", "announce", "bigrelease"]
    }).option("git", {
      describe: "Release as a git tag and push it",
      type: "boolean"
    }).option("npm", {
      describe: "Increment version in package.json",
      type: "boolean"
    }).option("package-json", {
      describe: "Increment version in package.json without publishing to npm",
      type: "boolean"
    }).option("push-main", {
      describe: "Push the main branch after releasing",
      type: "boolean"
    }).option("no-push-main", {
      describe: "Disable pushing to main branch (overrides --push-main)",
      type: "boolean"
    }).option("readme", {
      describe: "Increment version in README files",
      type: "boolean"
    });
  },
  async (argv) => {
    const ctx = await getAppContext({ argv });
    await release(ctx).catch(handleCommandError);
  }
).command(
  "stage [action]",
  "Tag the current commit and stage changes locally",
  (yargs2) => {
    return yargs2.positional("action", {
      describe: "The stage action to perform",
      default: "auto",
      choices: ["auto", "increment", "announce", "bigrelease"]
    }).option("git", {
      describe: "Stage as a git tag",
      type: "boolean"
    }).option("npm", {
      describe: "Stage version in package.json",
      type: "boolean"
    });
  },
  async (argv) => {
    const ctx = await getAppContext({ argv });
    await stage(ctx).catch(handleCommandError);
  }
).argv;
